pkg/pss/pss_test.go

```diff
@@ -146,23 +104,99 @@ func TestDeliver(t *testing.T) {
 	}
 }
 
-func TestHandler(t *testing.T) {
-	pss := pss.New(logging.New(ioutil.Discard, 0), nil)
-	testTopic := trojan.NewTopic("TEST_TOPIC")
+// TestRegister verifies that handler funcs are able to be registered correctly in pss
+func TestRegister(t *testing.T) {
+	var (
+		pss     = pss.New(logging.New(ioutil.Discard, 0))
+		h1Calls = 0
+		h2Calls = 0
+		h3Calls = 0
+		mtx     sync.Mutex
+
+		topic1  = trojan.NewTopic("one")
+		topic2  = trojan.NewTopic("two")
+		payload = []byte("payload")
+		target  = trojan.Target([]byte{1})
+		targets = trojan.Targets([]trojan.Target{target})
+
+		h1 = func(_ context.Context, m *trojan.Message) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			h1Calls++
+		}
+
+		h2 = func(_ context.Context, m *trojan.Message) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			h2Calls++
+		}
+
+		h3 = func(_ context.Context, m *trojan.Message) {
+			mtx.Lock()
+			defer mtx.Unlock()
+			h3Calls++
+		}
+	)
+	_ = pss.Register(topic1, h1)
+	_ = pss.Register(topic2, h2)
 
-	// verify handler is null
-	if pss.GetHandler(testTopic) != nil {
-		t.Errorf("handler should be null")
+	// send a message on topic1, check that only h1 is called
+	msg, err := trojan.NewMessage(topic1, payload)
+	if err != nil {
+		t.Fatal(err)
+	}
+	c, err := msg.Wrap(targets)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = pss.TryUnwrap(context.Background(), c)
+	if err != nil {
+		t.Fatal(err)
 	}
+	checkCalls(t, h1Calls, 1)
+	checkCalls(t, h2Calls, 0)
 
-	// register first handler
-	testHandler := func(ctx context.Context, m *trojan.Message) error { return nil }
+	// register another topic handler on the same topic
+	cleanup := pss.Register(topic1, h3)
+	err = pss.TryUnwrap(context.Background(), c)
+	if err != nil {
+		t.Fatal(err)
+	}
+	checkCalls(t, h1Calls, 2)
+	checkCalls(t, h2Calls, 0)
+	checkCalls(t, h3Calls, 1)
 
-	// set handler for test topic
-	pss.Register(testTopic, testHandler)
+	cleanup() // remove the last handler
```

Comment:
should this not be called with defer before fatal  calls after ?

Reply:
well not explicitly since `pss.Register` does not take any resources per-se, i.e. it does not start any goroutine for the handlers on registration, only on notification. That's why I omitted the cleanups on the other handlers